pkg/server/datastore/sqlstore/sqlstore.go

```diff
@@ -739,11 +739,49 @@ func (ds *Plugin) openDB(cfg *configuration, isReadOnly bool) (*gorm.DB, string,
 			db.Close()
 			return nil, "", false, nil, err
 		}
+		// TODO: we should keep this logic for a minor release cycle to make sure stale entries are removed eventually.
+		// Delete this function in a future release.
+		if err := cleanStaleNodeResolverEntries(db, ds.log); err != nil {
+			ds.log.WithError(err).Error("Failed to clean stale node resolver entries")
+		}
 	}
 
 	return db, version, supportsCTE, dialect, nil
 }
 
+func cleanStaleNodeResolverEntries(tx *gorm.DB, log logrus.FieldLogger) error {
```

Comment:
Instead of hydrating all of the ids for rows that need to be deleted, what do you think about doing something simple like:

```
DELETE FROM node_resolver_map_entries WHERE spiffe_id NOT IN (SELECT spiffe_id FROM attested_node_entries)
```

Reply:
Initially, I thought about using subquery like that, but from what I know, subquery tends to be less performant than joining tables. This query will hold spire initialization only for the first time, so I preferred to use a more performant query here

Reply:
I appreciate that you are being performance minded :) I wrote some code to build up a dataset and did an "EXPLAIN ANALYZE" and indeed your version is faster by a large margin, at least with Postgres.

Carry on!

Reply:
I've evaluated an Explain Analyse in both queries using some sample data, resulting in 667 stale entries being deleted, with 652 attested nodes and 1000 node resolver entries.

Using Join:
![image](https://user-images.githubusercontent.com/33766735/222591776-e8b5d1c6-00be-442b-a845-3b20997b7ade.png)

Using Subquery
![image](https://user-images.githubusercontent.com/33766735/222591632-cd332265-035c-4845-883d-1fc9dd933b0b.png)

Apparently my optimization attempt was unsuccessful, I'm going to update it :stuck_out_tongue:

Reply:
@guilhermocc and I got different results. We did some more analysis and the subquery version is indeed faster and less costly and has the benefits of not having to stash the row ids in memory on the server side before deleting them.